### PR TITLE
PRIME-2475 Backend process for reporting

### PR DESCRIPTION
**Pull request by bergomi02**

### CODE REVIEW

File: prime-dotnet-webapi/Controllers/JobsController.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        [HttpPost("populate/practitioner", Name = nameof(PopulatePractitionerTable))]
+        [Authorize(Roles = Roles.PrimeApiServiceAccount)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        public async Task<ActionResult> PopulatePractitionerTable()
+        {
+            await _reportingService.PopulatePractitionerTableAsync();
+            return NoContent();
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/JobsController.cs:39]: For debugging, should return some useful information, e.g. number of `Practitioner` rows added

File: prime-dotnet-webapi/Controllers/JobsController.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        [HttpPost("update/practitioner", Name = nameof(UpdatePractitionerTable))]
+        [Authorize(Roles = Roles.PrimeApiServiceAccount)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        public async Task<ActionResult> UpdatePractitionerTable()
+        {
+            await _reportingService.UpdatePractitionerTableAsync();
+            return NoContent();
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/JobsController.cs:54]: For debugging, should return some useful information, e.g. number of `Practitioner` rows updated

Comment (bergomi02) [reply prime-dotnet-webapi/Controllers/JobsController.cs]: Added return string to the API call.

File: prime-dotnet-webapi/Services/interfaces/IReportingService.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using System.Threading.Tasks;
+using System;
+
+namespace Prime.Services
+{
+    public interface IReportingService
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/interfaces/IReportingService.cs:6]: Can comments be added, possibly using text from PRIME-2475?

Comment (bergomi02) [reply prime-dotnet-webapi/Services/interfaces/IReportingService.cs:9]: comment added

File: prime-dotnet-webapi/Models/Practitioner.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Prime.Models
+{
+    /// <summary>store unautorized access college id and pracRefId from transaction log then fill the first name and last name from PharmaNetAPI.</summary>
+    [Table("Practitioner")]
+    public class Practitioner : BaseAuditable
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/Practitioner.cs:9]: Is `UnauthorizedPractitioner` more accurate?

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Prime.Models;
+using Prime.Models.Plr;
+using Prime.ViewModels.Plr;
+using Prime.ViewModels;
+using Prime.HttpClients;
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:14]: Lots of unused imports

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                    startDate = DateTime.Now.AddDays(-CalculationPeriodInDays);
+                }
+
+                if (endDate == null)
+                {
+                    endDate = DateTime.Now;
+                }
+
+                // get all approved enrollee
+                var enrolleeCerts = _context.Enrollees
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:43]: Even though the table is misnamed `Certifications`, we should keep "cert" out of the variable names:  `enrolleeLicences` instead?

**Review by neophyte57 — CHANGES_REQUESTED** [review]

File: prime-dotnet-webapi/Services/ReportingService.cs (hunk 6 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Services/ReportingService.cs]: variable name updated.

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                    .Where(e => e.GPID != null && e.Certifications.Any())
+                    .Select(e => new
+                    {
+                        e.Certifications.FirstOrDefault().LicenseNumber,
+                        e.Certifications.FirstOrDefault().Prefix
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:48]: `Certification.Prefix` may be `null`

Comment (bergomi02) [reply prime-dotnet-webapi/Services/ReportingService.cs:48]: Interesting observation. I did a quick query in Prod PharmaNet Translation log and there is no record where College Prefix is null and Practitioner ID is not null. 
Updated the query to return the ones Prefix is not null.

Comment (neophyte57) [reply prime-dotnet-webapi/Services/ReportingService.cs:48]: To clarify, should it be like:  https://github.com/bcgov/moh-prime/blob/c12e19a5d4480bce8660fbc7a1f6318fffab1eb6/prime-dotnet-webapi/Services/EnrolleeService.cs#L1072C70-L1072C70

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                {
+                    endDate = DateTime.Now;
+                }
+
+                // get all approved enrollee
+                var enrolleeLicences = _context.Enrollees
+                    .Where(e => e.GPID != null && e.Certifications.Any(c => c.Prefix != null))
+                    .Select(e => new
+                    {
+                        e.Certifications.FirstOrDefault().LicenseNumber,
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:47]: For nurses and midwives, I think `e.Certifications.FirstOrDefault().PractitionerId` is what we should compare against to `PharmanetTransactionLog.PractitionerId`

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                    _context.Update(p);
+                }
+                result = $"practitionerRecords count = {practitionerRecords.Count}";
+
+                await _context.SaveChangesAsync();
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Error: PopulatePractitionerTableAsync - Message: {ex.Message}");
+                result = $"Error: PopulatePractitionerTableAsync - Message: {ex.Message}";
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:113]: Use `{nameof(PopulatePractitionerTableAsync)}` instead

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+                result = $"Start date = {startDate:dd MMM yyyy}, End Date = {endDate:dd MMM yyyy}, questionablePractitionerIds count = {questionablePractitionerIds.Count} ";
+
+                await _context.SaveChangesAsync();
+
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Error: PopulatePractitionerTableAsync - Message: {ex.Message}");
+                result = $"Error: PopulatePractitionerTableAsync - Message: {ex.Message}";
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:80]: Use {nameof(PopulatePractitionerTableAsync)} instead

File: prime-dotnet-webapi/Services/ReportingService.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                    }).ToListAsync();
+
+                foreach (var p in questionablePractitionerIds)
+                {
+                    _context.Practitioner.Add(new Practitioner
+                    {
+                        CollegeId = p.PractitionerId,
+                        PracRefId = p.CollegePrefix
+                    });
+                }
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/ReportingService.cs:76]: If there are multiple records in `PharmanetTransactionLog` for the same `PractitionerId`/`CollegePrefix` pair, do we want multiple `Practitioner` records created?

Comment (bergomi02) [reply prime-dotnet-webapi/Services/ReportingService.cs:76]: Good catch. it is fixed.

**Review by neophyte57 — APPROVED** [review]

**Pull request merged**